org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/BuildCycleManager.xtend

```diff
@@ -39,20 +42,57 @@ class BuildCycleManager {
 	var IndexState lastIndexState = new IndexState
 	var String[] staticSearchPaths = null
 
+	var ChangedResources currentChanges
+
 	def void startBuild() {
-		createBuildRequest.addChanges.build.updateIndex
-		updateValidationMarkers
+		val initialBuild = resourceDescriptionsProvider.data === null
+
+		val buildRequest = createBuildRequest.addChanges
+
+		if (initialBuild) {
+			buildRequest.build => [
+				indexState.updateIndex
+				buildRequest.state = indexState
+			]
+		}
+
+		buildRequest => [
+			build => [
+				indexState.updateIndex
+				affectedResources.filter[getNew !== null].map[uri].forEach[validate]
+				affectedResources.filter[getNew === null].map[uri].forEach[removeValidationMarkers]
```

Comment:
why is this step necessary. It probably has something to do with the initial build, right?

Reply:
I extracted the validation out of the incremental builder, although this might not be necessary anymore. I can try to return this to using the incremental builder's internal validation invocation and the `afterValidate` callback.